src/hivpy/hiv_testing.py

```diff
@@ -145,67 +194,77 @@ def update_anc_hiv_testing(self, pop, time_step):
-                r = rng.uniform(size=len(tested_uncirc_male_pop))
-                circumcision = r < pop.circumcision.prob_circ_after_test
-                # assign outcomes
-                pop.set_present_variable(col.CIRCUMCISED, circumcision, tested_uncirc_male_pop)
-                pop.set_present_variable(col.VMMC, circumcision, tested_uncirc_male_pop)
-
-    def update_post_vmmc_testing(self, pop):
-        """
-        Update HIV testing status after VMMC.
+            self.update_sub_pop_test_mark(pop, post_delivery_pop, self.prob_test_postdel)
+
+    def update_sub_pop_test_mark(self, pop, sub_pop, prob_test):
         """
-        # those that just got circumcised and weren't tested last time step get tested now
-        just_tested = pop.get_sub_pop(AND(COND(col.CIRCUMCISION_DATE, op.eq, pop.date),
-                                          OR(COND(col.LAST_TEST_DATE, op.lt, pop.date - timedelta(days=90)),
-                                             COND(col.LAST_TEST_DATE, op.eq, None))))
-        # correctly set up related columns
-        if len(just_tested) > 0:
-            pop.set_present_variable(col.EVER_TESTED, True, just_tested)
-            pop.set_present_variable(col.LAST_TEST_DATE, pop.date, just_tested)
-            # "reset" dummy partner columns
-            pop.set_present_variable(col.NSTP_LAST_TEST, 0, just_tested)
-            pop.set_present_variable(col.NP_LAST_TEST, 0, just_tested)
+        Update the test mark of a sub-population based on a given probability.
+        """
+        if len(sub_pop) > 0:
+            # mark people for testing
+            r = rng.uniform(size=len(sub_pop))
+            marked = r < prob_test
+            # set outcomes
+            pop.set_present_variable(col.TEST_MARK, True, sub_pop=pop.apply_bool_mask(marked, sub_pop))
+
+    def test_mark_general_pop(self, pop):
+        """
+        Mark general population to undergo testing this time step.
+        """
+        # testing occurs after a certain year if there is no covid disruption
+        if ((pop.date.year >= self.date_start_testing)
+           & (not (self.covid_disrup_affected | self.testing_disrup_covid))):
+
+            # update testing probabilities
+            self.rate_first_test = self.init_rate_first_test + (min(pop.date.year, self.date_test_rate_plateau)
+                                                                - self.date_start_testing) \
+                                                                * self.an_lin_incr_test
+            self.rate_rep_test = (min(pop.date.year, self.date_test_rate_plateau)
+                                  - self.date_start_testing) * self.an_lin_incr_test
+
+            # general population ready for testing
+            testing_population = pop.get_sub_pop([(col.HARD_REACH, op.eq, False),
+                                                  (col.AGE, op.ge, 15),
+                                                  (col.HIV_STATUS, op.eq, False),
```

Comment:
```suggestion
                                                  (col.HIV_DIAGNOSED, op.eq, False),
```
needs to be diagnosed, not HIV status, because otherwise no HIV+ people will get tested!